### DISCUSSION TITLE
How to lazy load surround?

### DISCUSSION THREAD

**Discussion by AckslD** [wbthomason/packer.nvim#487] [Q&A]
I'm trying to lazy load [vim-sandwich](https://github.com/machakann/vim-sandwich):
```lua
use {
    'machakann/vim-sandwich',
    keys = {'sa', 'sd', 'sr', 'sdb', 'srb'},
}
```

but the first time I press for example `sa` it doesn't work. After that it works, any idea how to solve this?

**Comment by wbthomason** [comment]
Hmm. I can't quite confirm this - when I try to replicate your example, the `sandwich` mappings load correctly on first use. Can you post the relevant part of your compiled file?

**Reply by wbthomason** [reply]
I will also point out that - while you're of course welcome to lazy-load whatever you need to - on my system, the startup impact of eagerly-loaded `sandwich` is minimal (>1ms).

**Reply by AckslD** [reply]
Thanks @wbthomason , I'll share the compiled part after the weekend. And good to know that lazy loading might not be so important here 👍

**Reply by AckslD** [reply]
@wbthomason here's the compiled `sandwich` part:
```lua
time([[try_loadstring definition]], false)
time([[Defining packer_plugins]], true)
_G.packer_plugins = {
  ...
  ["vim-sandwich"] = {
    keys = { { "", "sa" }, { "", "sd" }, { "", "sr" }, { "", "sdb" }, { "", "srb" } },
    loaded = false,
    needs_bufread = true,
    path = "/home/axel/.local/share/nvim/site/pack/packer/opt/vim-sandwich"
  },
  ...
}
-- Keymap lazy-loads
time([[Defining lazy-load keymaps]], true)
...
vim.cmd [[noremap <silent> sa <cmd>lua require("packer.load")({'vim-sandwich'}, { keys = "sa", prefix = "" }, _G.packer_plugins)<cr>]]
...
time([[Defining lazy-load keymaps]], false)
```

I wonder it there is some clash wish `which-key` here similar to https://github.com/folke/which-key.nvim/issues/130 as after starting I have the following mappings (from `:nmap s`):
```
n  sÞ          * <Nop>
n  s           * <Cmd>lua require("which-key").show("s", {mode = "n", auto = true})<CR>
   sr          * <Cmd>lua require("packer.load")({'vim-sandwich'}, { keys = "sr", prefix = "" }, _G.packer_plugins)<CR>
   sdb         * <Cmd>lua require("packer.load")({'vim-sandwich'}, { keys = "sdb", prefix = "" }, _G.packer_plugins)<CR>
   srb         * <Cmd>lua require("packer.load")({'vim-sandwich'}, { keys = "srb", prefix = "" }, _G.packer_plugins)<CR>
   sd          * <Cmd>lua require("packer.load")({'vim-sandwich'}, { keys = "sd", prefix = "" }, _G.packer_plugins)<CR>
   sa          * <Cmd>lua require("packer.load")({'vim-sandwich'}, { keys = "sa", prefix = "" }, _G.packer_plugins)<CR>
```

**Reply by AckslD** [reply]
I can confirm that adding `ignore_missing = true` to which-key config indeed fixes this issue

**Reply by AckslD** [reply]
Actually, even with `ignore_missing = true`, it only sometimes work. Does the compiled part look alright to you @wbthomason?

**Reply by AckslD** [reply]
I also tried:
```lua
                triggers_blacklist = {
                    -- list of mode / prefixes that should never be hooked by WhichKey
                    -- this is mostly relevant for key maps that start with a native binding
                    -- most people should not need to change this
                    n = {"s"},
                    i = {"j", "k"},
                    v = {"j", "k", "s"},
                    x = {"s"},
                },
```

in config for which-key.nvim so not have any other keybinds starting with `s` but also does not work.

So `:map s` now gives initially:
```
   sdb         * <Cmd>lua require("packer.load")({'vim-sandwich'}, { keys = "sdb", prefix = "" }, _G.packer_plugins)<CR>
   sa          * <Cmd>lua require("packer.load")({'vim-sandwich'}, { keys = "sa", prefix = "" }, _G.packer_plugins)<CR>
   srb         * <Cmd>lua require("packer.load")({'vim-sandwich'}, { keys = "srb", prefix = "" }, _G.packer_plugins)<CR>
   sr          * <Cmd>lua require("packer.load")({'vim-sandwich'}, { keys = "sr", prefix = "" }, _G.packer_plugins)<CR>
   sd          * <Cmd>lua require("packer.load")({'vim-sandwich'}, { keys = "sd", prefix = "" }, _G.packer_plugins)<CR>
```
and after pressing `sd` once (which fails):
```
n  srb           <Plug>(operator-sandwich-replace)<Plug>(operator-sandwich-release-count)<Plug>(textobj-sandwich-auto-a)
n  sdb           <Plug>(operator-sandwich-delete)<Plug>(operator-sandwich-release-count)<Plug>(textobj-sandwich-auto-a)
n  sr            <Plug>(operator-sandwich-replace)<Plug>(operator-sandwich-release-count)<Plug>(textobj-sandwich-query-a)
n  sd            <Plug>(operator-sandwich-delete)<Plug>(operator-sandwich-release-count)<Plug>(textobj-sandwich-query-a)
x  sr            <Plug>(operator-sandwich-replace)
x  sd            <Plug>(operator-sandwich-delete)
o  sa            <Plug>(operator-sandwich-g@)
x  sa            <Plug>(operator-sandwich-add)
n  sa            <Plug>(operator-sandwich-add)
```

**Reply by wbthomason** [reply]
Hmm, ok. This looks right, and we're clearing the "fake" keybind correctly. It seems like we're not "replaying" the keys correctly, which should be happening here: https://github.com/wbthomason/packer.nvim/blob/master/lua/packer/load.lua#L78-L102. I'm not sure why not, though. I'll try to reproduce it on my end.

**Reply by wbthomason** [reply]
@AckslD I can't reproduce this - `sandwich` seems to work as expected for me even when lazy-loaded like this.

**Reply by AckslD** [reply]
Thanks for trying @wbthomason! Any idea of what I might try to debug or figure out what's going wrong?

**Reply by wbthomason** [reply]
One thing I did notice is that, when I only ran `PackerCompile` when switching from eagerly-loaded to lazy-loaded, the keymaps did not work as intended because compiling doesn't move from start to opt. Make sure you've run a command that does that, like `PackerSync` or `PackerUpdate`. 

Although the keymaps still seem to work, there is a weird temporary "flash" of another buffer the first time they're activated for me, so we might still not be replaying the keys correctly. You could try modifying the code block in `packer` that I linked above to print the keys that we give to `feedkeys`, which might tell you if something unexpected is happening? You could also run `vim.cmd("verbose map sd")` or similar in that block to ensure that the `sandwich` bindings are being correctly created.

**Reply by AckslD** [reply]
I'll try this and let you know if I find something 👍

**Reply by AckslD** [reply]
@wbthomason I think I found what's going on! I noticed that when I pressed the keys the first time the cursor jumped to where the cursor was when I opened the file. This seems to happen since I have the following autocmd in my config which puts the cursor where it was last time when I open a new file:
```
" return to last edit position when opening files
autocmd BufReadPost * if line("'\"") > 1 && line("'\"") <= line("$") | exe "normal! g'\"" | endif
```

Not sure why `BufReadPost is triggered though

**Reply by wbthomason** [reply]
Ah, interesting. Thanks for continuing to chase this down. We have some logic (https://github.com/wbthomason/packer.nvim/blob/master/lua/packer/compile.lua#L248-L257) to detect if a plugin will need `BufRead` triggered to load properly - maybe that happens for `sandwich`?

**Reply by AckslD** [reply]
@wbthomason Indeed `sandwich` has a `after/ftplugin` folder. I have to think how to still be able to jump to last position when opening a new file.

**Reply by wbthomason** [reply]
It's definitely not my place to tell you how to write your config, but I'd personally just opt to eager-load `sandwich`. It's well-designed and doesn't have a high startup cost (for me, at least), and that way your existing autocmd works as you want.

**Reply by AckslD** [reply]
That's a fair point :)
